### ISSUE TITLE
`useTranslation` returns an unstable reference to `t`

### ISSUE THREAD

**Issue by eomuraliev** [i18next/react-i18next#1228]
I tried searching for mentions of unstable reference to `t` returned by `useTranslation` and did not find anything. Also, the docs do not mention whether the reference is expected to be stable or not.

-->

## 🐛 Bug Report

`useTranslation` returns an unstable reference of `t`, this causes problems when using it on other hooks.

## To Reproduce

```js
const { t } = useTranslation('fakespace');
// when not ready, `t` is always a different function every time `useTranslation` is invoked.
```

## Expected behavior

`t` should return a stable reference if arguments to `useTranslation` have not changed.

## Your Environment

- *runtime version*: i.e. node v14, webpack, chrome
- *i18next version*: i.e. 19.8.4
- *os*: Mac

**Comment by tigerabrodi** [comment]
@eomuraliev If it should be memoized or not is a question the maintainers can answer, what I am wondering is, what are the problem when using it with other hooks? Are we talking about performance problems or? :thinking:

Thanks :two_hearts: :pray:

**Comment by eomuraliev** [comment]
@tigerabrodi The problem is for example if I'm using `t` in a `useEffect` to show a user a localized notification, I need to make sure I have the latest and correct reference to `t` in the `useEffect` hook, so I want to include `t` in its dependencies. If `t` is not stable, then it will cause my `useEffect` to needlessly rerun all the time.

**Comment by eomuraliev** [comment]
I removed my guesses for where the problem is caused or how to solve it. I could be wrong on the causes, and I don't want it to distract from the discussion that `t` should be stable.

**Comment by tigerabrodi** [comment]
@eomuraliev Gotcha! Seems like memoziation would be the appropriate solution in my eyes as well, thanks for elaborating also! :tada:

**Comment by pedrodurek** [comment]
Hey guys, `t` function it's a state, so it wouldn't need to be memorized: https://github.com/i18next/react-i18next/blob/master/src/useTranslation.js#L42. I think it may be something else. 🤔

**Comment by eomuraliev** [comment]
After some digging. I think the problem in my case is that the value returned for `notReady` is not stable. The `useState` portion is *probably* stable.

**Comment by pedrodurek** [comment]
Do you mean `ready` property? If so, it's a boolean, it won't get a new "reference" on every rerender.

**Comment by eomuraliev** [comment]
I mean the values returned when `ready` is false. https://github.com/i18next/react-i18next/blob/8fa7101bc0cb1c39a9ef9b9a47c9f497af08ee21/src/useTranslation.js#L11-L19

`notReady` gets created to be assigned to `t` every time the hook is called https://github.com/i18next/react-i18next/blob/8fa7101bc0cb1c39a9ef9b9a47c9f497af08ee21/src/useTranslation.js#L13

**Comment by tigerabrodi** [comment]
Seems like wrapping it in a useCallback could help? Then it would only be redeclared whenever k changes right?

**Comment by eomuraliev** [comment]
`k` is an argument, so you could get away with just wrapping it in a `useCallback with an empty dependencies array. Also, you would need to lift the declaration of the `notReady` function up out of the `if`, since hooks should not be conditionally invoked.

**Comment by pedrodurek** [comment]
Isn't mandatory to pass `i18n` instance to `react-i18next`?

**Comment by tigerabrodi** [comment]
Yeah, this one is a bit tricky, one thing is for sure though, we wanna memoize it haha :smile:

**Comment by jamuhl** [comment]
@eomuraliev if you end in `if (!i18n)` it actually tells you - your config/setup is broken!!! you should never get into that execution path - it only exists to warn you that you got something wrong.

`t` function gets only set if something changes, same for the ready boolean (which as a boolean is a type and not a ref)

**Comment by akamfoad** [comment]
I'm curious if there is any plan/progress on this issue, the only workaround I came up with was totally remove `t` from dependency array of hooks like `useEffect` (it triggers API calls in my case) , don't know if that has any negative consequences.

**Comment by jamuhl** [comment]
@akamfoad `t` is taken from state: https://github.com/i18next/react-i18next/blob/master/src/useTranslation.js#L42 -> it's replaced if either translations were loaded or language was changed -> which makes sense

so, not 100% sure what your issue is? But regarding API calls, I guess those do not depend on `t` so it should not be in dependency array

**Comment by akamfoad** [comment]
> `t` is taken from state: https://github.com/i18next/react-i18next/blob/master/src/useTranslation.js#L42 -> it's replaced if either translations were loaded or language was changed -> which makes sense

I thought it could be stable between renders cuz it seems like same `t` function which is independent of translations and localization boilerplate, it takes a key (or keys and options) and returns a string (usually), don't know if that make sense.

> so, not 100% sure what your issue is? But regarding API calls, I guess those do not depend on t so it should not be in dependency array

You're right, but I have a use case like below
```js
useEffect(()=>{
  const getSMTH = async () => {
    try{
      const res = await fetch("...");
      setData(await res.json())
    } catch(err){
      setPopupText(t("Error_Key"))
    }
  getSMTH()
}, [aParam, anotherParam, t, setPopupText]);
```

Is there any pattern to the use cases you can think of then?

**Comment by jamuhl** [comment]
no - the `t` function is bound to the namespace set on useTranslation and acts as a rerender trigger prop for the case something changed and i18n needs rerendering. Every other case the `t` function is stable.

You got the `t` function mentioned in the issue -> than something is wrong in your code.

Regarding API call passing t -> you can use i18n.t for that case...or set an error state and call t on render

**Comment by akamfoad** [comment]
> You got the t function mentioned in the issue -> than something is wrong in your code.

Nah sorry about that, I didn't mean to get help with my code hahaha.
What I meant was, I thought `t` is a function which takes an argument and depending on some global state (i18n object and its resources **itc**) returns a string, but seems it is more complicated than that.

I think the example was too small to make it obvs 😬 
Thanks for the clarification @jamuhl

**Comment by MerlinMason** [comment]
Just noticed this same problem on our project, a useEffect hook was constantly firing.

Switching from
```js
const lang = useTranslation();
...
lang.t("foo");
```
to
```js
const { t } = useTranslation();
...
t("foo");
```
solved the issue for us.

**Comment by thiagoevg** [comment]
> > `t` is taken from state: https://github.com/i18next/react-i18next/blob/master/src/useTranslation.js#L42 -> it's replaced if either translations were loaded or language was changed -> which makes sense
> 
> I thought it could be stable between renders cuz it seems like same `t` function which is independent of translations and localization boilerplate, it takes a key (or keys and options) and returns a string (usually), don't know if that make sense.
> 
> > so, not 100% sure what your issue is? But regarding API calls, I guess those do not depend on t so it should not be in dependency array
> 
> You're right, but I have a use case like below
> 
> ```js
> useEffect(()=>{
>   const getSMTH = async () => {
>     try{
>       const res = await fetch("...");
>       setData(await res.json())
>     } catch(err){
>       setPopupText(t("Error_Key"))
>     }
>   }
>   getSMTH()
> }, [aParam, anotherParam, t, setPopupText]);
> ```
> 
> Is there any pattern to the use cases you can think of then?

I'm having exactly the same issue. =/

**Comment by jamuhl** [comment]
@thiagoevg 

do you really have an issue? or is there just something in your code that is sub-optimal?

looking at the problem and possible solution:

a) make `t` "stable" not changing if language changes or translations are loaded -> result your popup is filled with either no translations or wrong language depending on what triggered the change of `t`

b) not using `t` in the dependency array (which should not be named dependency array but props array -> as only those are needed to set there which trigger a rerender on change) -> same result as a) your text in the popup is wrong

c) don't set the popup text  but `setPopupKey("Error_Key")` -> and do based on that error key the translations inside the rendering of your popup

as a general rule -> keep translation as close as possible to the rendering and as far as possible from API calls

**Comment by kitsunekyo** [comment]
we ran into the same issue with an effect.

our `useEffect` uses `t` to show a toast message, and as soon as i add `t` into the dependency array, the effect runs every render. i would also have expected `t` to be stable. or at least have it documented somewhere whether it actually is or not. moving the resolving of translations into somewhere else is not an option. i might not always want to pass translation keys.

i reckon this isnt an exotic usecase. so it would be great to have an answer from a maintainer.

**Comment by jamuhl** [comment]
`t` is only changed if needs to be changed https://github.com/i18next/react-i18next/blob/master/src/useTranslation.js#L65

-> when language changed, finished loading something, ....

**Comment by git-angelo** [comment]
I did not find how to solve the problem but here we can see the problem with the less code possible: https://codesandbox.io/s/snowy-hill-vwljzl?file=/src/index.js

Open the terminal of the codesanbox and you'll see:
- 2 calls without "t" as dependency.
- 3 calls with "t" as dependency.
- 2 calls with "i18n"

As @kitsunekyo , I use t() to display a toast in useEffect and I have 3 API calls triggered (2 from React, 1 from "t" as dependency) and for now the only solution that I found is to use i18n.t() but it does not return the exact same result as t() and so it's not ideal (for me).

NB: The first two renders are due to React 18: https://reactjs.org/docs/strict-mode.html#ensuring-reusable-state
